### PR TITLE
Flood segmentation competition benchmark for PC

### PR DESCRIPTION
**Pull request by tamara-glazer**
The benchmark solution for the challenge [STAC Overflow: Map Floodwater from Radar Imagery](https://www.drivendata.org/competitions/81/detect-flood-water/) has been updated to run end-to-end on the Planetary Computer. This solution makes the following updates to the original solution:

- pip installations for packages not available on the PC (commented out by default)
- `self.save_hyperparameters()` removed from the `FloodModel`
- `filepath` column removed from the training metadata; instead, `FloodModel` now takes a feature directory and label directory as input via `hparams`; a `_build_df` helper method is added to the `FloodModel` class to instantiate training and validation dataset objects using these directories
- updates made to the tutorial text explaining where to save the data files

### CODE REVIEW

File: competitions/s1floods/benchmark_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,2029 @@
...
+   "metadata": {},
+   "source": [
+    "### Metadata"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First we'll download the training data to a new subdirectory called `training_data/` in our current working directory. If you are working on the Planetary Computer Hub, consider using [Azure Blob Storage](https://docs.microsoft.com/en-us/azure/storage/blobs/storage-blobs-introduction) to store your data using the West Europe region, which is where your server is running.\n",
```
Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:118]: Can you add a link to https://planetarycomputer.microsoft.com/docs/quickstarts/storage/ as well. Something like

> See [Writing data to Azure Blob Storage](https://planetarycomputer.microsoft.com/docs/quickstarts/storage/) for an example.

File: competitions/s1floods/benchmark_tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,2029 @@
...
+    "    ├── awc00.tif\n",
+    "    ├── awc01.tif\n",
+    "        ...\n",
+    "    ├── wvy30.tif\n",
+    "    └── wvy31.tif\n",
+    "\n",
+    "2 directories, 1627 files\n",
+    "```\n",
+    "\n",
+    "To better understand what's in our data, we can begin by exploring trends across different locations and flood events. Let's load the metadata and look at what we have."
```
Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:140]: Can you merge the next few cells, just so that this is easier to run through in fewer clicks?

File: competitions/s1floods/benchmark_tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,2029 @@
...
+    "flood_counts.describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The training data include chips from 13 flood events. We have anywhere from 15 to 69 chips (30 to 138 images) per unique event, with half of events containing fewer than 32 chips (64 images).\n",
+    "\n",
+    "We can also take a look at the distribution of chips by location."
```
Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:363]: We should have seaborn installed, in which case the next cell is something like `sns.countplot(x="location", y="chip_id")` (untested).

File: competitions/s1floods/benchmark_tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,2029 @@
...
+    "**Recall that each set of two polarizations (vh and vh) correspond with a single water label.**"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import rasterio"
```
Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:516]: Again, a few cells that can be merged (up until printing `metadata`)

File: competitions/s1floods/benchmark_tutorial.ipynb [hunk 5]
```diff
@@ -0,0 +1,2029 @@
...
...KiUg058i5YtUu2egXtmVX2BavehlnK6UmBfWMZSzsl6Z/gDnLchkaSf7lV2QtWqWTtDuXxPcJzq/T9XZyBzGx1GXtVpdXVsI4f0e7un1l1vExlccKy+J9/zqo/o26PGHsrPNx9m8G5VTcHH7JV/JBO5zK+9QoffXwLbx6gDbfRP0l4vXXBqk9XOH57mSX7XbqlUw6bvQtWfQvn6Nyq/NwqnltV5BLGuVXCC1aFSElAoRnIiwHDxQoFIuryU6se/Fy9/KUYMpIms959TH1lxtC5Siv/GMvYxU8aCKUm84nJYlKhZTnoWggbPfJqmdBVKQ1uY3j7uOkGXsOk7KdUDj+hZRwy33yVRe9tFO2APKsQRyJFMKau3yfIVCqTOzj211nYL6GUR0jTCXa6y+SJRBYaZDWTPsvIsoosVXD1gq5wnyI+RNIEpvYK1ZbPfNfH8B4j5QaDWMFOR2hbBhNphem8wpX2EywhwTuuE/QuIzZXcYoutfge1tEnpJKF2CojpT7uIsXK92mMHxMWJXQtxK9cot7NkCcPsBWHxFhiJ/oNUrVNCZ+a95BMhjwoUPwpg9IGQlaQaDpir8qwaCCIOYUYMPKqCEGIef82QsuC7u8ivlVi4/6Y7cMWn+Ut3O0xb1/6GV9X/iUf7Hwda03A0mr0t95hNFhitfYIL19iP+pQaBly5CEHGRVxD1nTUdWY+H6I0K3ilzo0zD7TtMdoHlPsjGmuDPCWN5iXNsjjEdpSirG/i1KRKUljMtVElhWkOCecFPhCiBuapJUGq+1T4u0jDj4qSEOLQNJZffQuCTaqXiDVUyRFoFhEzAKdQhBx8yXK97bpTO9Dq8S8qOJIXZTFDM3MqLoPmI6uQT1gadNFD4458bYIvGvIR1NuKH9MNB1xEn+Vq/M/5JXWbfSVlLtpje3TLSzvAEGok88WzBbw8LM2an2Kv3uIF5aIpxohCnvRForoUBN95MSl399kY2UfVS+wpxPsaszIW2deew139zHldQ8/MpEsFS8roxPQCEb4lS7i0CUsoF56Qik5ZixvMa7cYB6WsYfbSO6CMHBRWzrJoxOSeYAvln7RVf+Vz5lV29RX5gydK2dW6bt4aQPBajKfmSzGF6za7JJXK4SOjDW4jeEdfM6qOzSNI2abr+H0rqGoBxRZhSQSyYMRdf0eQapRmd7GKX8Dx76JEo6QZhPseI
```
Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:960]: Would https://github.com/microsoft/PlanetaryComputerExamples/blob/main/competitions/s1floods/generate_auxiliary_input.ipynb notebook here? I think a relative link like

    See [Fetching supplementary model input](generate_auxiliary_input.ipynb) for an example.

will work.

Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:1171]: Can you post the output of this cell running on the Planetary Computer?

Previously, `pip installing` some additional package that depended on torch would wipe out our existing (conda-based. I've since upgraded the package versions, but haven't tried that again.

If this causes issues, we might `!mamba install -y -c conda-forge albumentations` instead.

Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:1302]: Same question here about the output of this command.

Comment (TomAugspurger) [inline competitions/s1floods/benchmark_tutorial.ipynb:1593]: We include [jupyter-server-proxy](https://jupyter-server-proxy.readthedocs.io/en/latest/) so I suspect that you can view tensorboard while it's running on the Hub.

If the URL of your notebook is

    https://pccompute.westeurope.cloudapp.azure.com/compute/user/taugspurger@microsoft.com/lab/tree/path/to/notebook.ipynb

then the URL of tensorboard would be

    https://pccompute.westeurope.cloudapp.azure.com/compute/user/taugspurger@microsoft.com/proxy/6006

So can you check that that works and update the line about needing to download the directory locally?

**Review by TomAugspurger — COMMENTED** [review]
Thanks, this looks great! I had a couple small style nitpicks, but I think the most important comments were:

1. Requests to add links to other notebooks / examples
2. Confirming that pip installing those libraries with torch deps doesn't reinstall all of pytorch

File: competitions/s1floods/benchmark_tutorial.ipynb (hunk 5 above)
Comment (tamara-glazer) [reply competitions/s1floods/benchmark_tutorial.ipynb:1087]: I haven't encountered any issues installing `albumentations`. I'll post the output of these pip installations so you can take a look to ensure we aren't wiping anything unexpected.

Comment (tamara-glazer) [reply competitions/s1floods/benchmark_tutorial.ipynb:1481]: I'm currently getting a "500: Internal Server Error" when I try 
https://pccompute.westeurope.cloudapp.azure.com/compute/user/tammy@drivendata.co/proxy/6006

Is this the correct URL?

**Comment by tamara-glazer** [comment]
@TomAugspurger Thanks for the great feedback. I've pushed an updated version, which includes all cell output for inspection (including pip installations). The only outstanding item is figuring out the best way to view the tensorboard.

**Comment by TomAugspurger** [comment]
Great thanks. If you get stuck on the tensorboard proxying stuff let me know and I can try it out.

**Comment by TomAugspurger** [comment]
FYI, https://github.com/microsoft/PlanetaryComputerExamples/blob/main/CONTRIBUTING.md has some notes on setting up pre-commit to fix the linting issue.

**Comment by TomAugspurger** [comment]
@tamara-glazer I had a chance to go through this today. Looks great.

One question about tensorboard: Does pytorch-lightning automatically start up a tensorboard server? Or do you need to open a terminal and run `tensorboard --log-dir=tensorboard_logs`?

How long should I expect between making a request to https://pcc-staging.westeurope.cloudapp.azure.com/compute/user/<username>/proxy/6006/  and having the UI appear? It's take ~1 minute for  me. Looking at the inspector, most of the time is spent downloading the 7.6 MB `index.js` file.

I'm going to push a commit fixing the formatting issues and mentioning how to start a tensorboard server.

**Comment by TomAugspurger** [comment]
CI is passing now, so I'll go ahead and merge this. @tamara-glazer let me know if the bit I added about starting a tensorboard server in https://github.com/microsoft/PlanetaryComputerExamples/pull/106/commits/bbbf1e4e4dda0da7c04af24f1d44a205724f4b0b is unnecessary, and I'll make another PR to remove it.

Thanks!

**Comment by tamara-glazer** [comment]
@TomAugspurger Thanks for adding the tensorboard text and for running the pre-commit. Typically I've always interacted with tensorboard by running `tensorboard --log-dir=tensorboard_logs` from the terminal like you describe. It should load very quickly, so any significant delays are unusual. The text you added is helpful and can definitely stay in. Thanks again for your review and for merging. Very excited to have this example shared on the PC!
